rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,446 @@
+
+However, it's possible to have some failure on flushing WAL to object storage which last for several minutes or even longer, and without a local disk which might help to buffer the WAL during the failure.
+
+In this case, we can still buffer the WAL in memory as long as possible. When the failure is resolved, we can resume the flush operation.
+
+However, the memory is limited, and we can't buffer the WAL in memory forever. We already have a threshold setting `max_unflushed_bytes` for this. When this threshold is reached, we can't buffer the WAL anymore, and we have to mark the db into an read-only state.
+
+When the WAL fails to be flushed to storage, the write operation with `SyncLevel::Remote` should be failed with an `IOError`, and considered as not committed. This write operation is ok to be visible to readers with `ReadWatermark::LastCommitting`, but should be invisible to readers with read watermark higher than `ReadWatermark::LastCommitted`.
+
+### Possible Code Changes
+
+As above described, we need to tackle with several details to implement this proposal, like:
+
+1. The synchronization of flushing the WAL to storage.
+2. Notify the writers when the WAL is successfully flushed to storage.
+3. Retry the flush operation when it fails with backoff, and finally mark the db into an read-only state if it's permanent failure.
+4. Flush the memory buffered WAL to storage in order.
+5. Support tiered WAL storage like local disk and S3.
+6. Handle the no-WAL mode.
+
+Given the complexity of these changes, implementing everything in a single PR would be challenging. It's difficult to estimate the implementation effort required or identify all the code that will need to change in this RFC. A better approach would be to break this down into several smaller PRs, starting with some small refactoring to make the subsequent changes easier to be implemented.
+
+One possible code change would be introducing a `WALManager` struct to centralize management of the WAL buffer and flush operations, helping to encapsulate the complexity in a single place.
+
+For inspiration, we can also consider to reference how Badger structures their code. In Badger's implementation, they embed the WAL buffer inside the memtable struct, as shown here:
+
+```go
+// memTable structure stores a skiplist and a corresponding WAL. Writes to memTable are written
+// both to the WAL and the skiplist. On a crash, the WAL is replayed to bring the skiplist back to
+// its pre-crash form.
+type memTable struct {
+	sl         *skl.Skiplist
+	wal        *logFile
+	maxVersion uint64
+	opt        Options
+	buf        *bytes.Buffer
+}
+```
+
+It makes sense to put the manager of the WAL buffer inside the memtable struct, as WAL is closely related to the memtable, it actually make a good encapsulation: put the complexities behind some simple `put()` / `get()` interface.
```

Comment:
I actually like that the memtable and wal are decoupled and that The WAL is a top-level object because its critical to how we do db fencing, which is quite unique to SlateDB. My preference would be to keep them as separate entities.

Reply:
in LSM engines, the WAL only needs to be concerned with data related to the Memtable, and the two are logically quite closely tied together. however, in terms of code organization, the WAL and Memtable can be separated. i think we could try to avoid making modifications on this unless we encounter difficulties during the implementation.